### ISSUE TITLE
glslify as build step

### ISSUE THREAD

**Issue by jsonnull** [mattdesl/gl-vignette-background#2]
I was surprised to learn that this module was pushing a couple hundred kilobytes added onto my build.

After a quick glance at the source, it looks like all of glslify is pulled in at runtime for this module. Is that right?

Shouldn't glslify—like browserify—be something run at build time?

I think for this repo, this means adding the build config and ensuring the necessary build files are published for users pulling in the module—OR, pushing the build requirement onto the end user.

**Comment by mattdesl** [comment]
Hmm what are you using for your build step? Browserify? Webpack?

It should inline the shaders as a string so there should be no extra kb from glslify. Maybe there is a bug tho. 😀

**Comment by jsonnull** [comment]
I'm using webpack.

Here's a quick before/after just importing the module, no other changes:

```
       Asset     Size  Chunks             Chunk Names
    build.js  23.4 kB       0  [emitted]  main
build.js.map  27.9 kB       0  [emitted]  main
    + 13 hidden modules

       Asset    Size  Chunks             Chunk Names
    build.js  215 kB       0  [emitted]  main
build.js.map  262 kB       0  [emitted]  main
    + 63 hidden modules
```

Since glslify is called directly from the module (https://github.com/mattdesl/gl-vignette-background/blob/master/index.js#L8-L9), I think glslify + deps are all entering the webpack build as those hidden dependencies.

**Comment by mattdesl** [comment]
Try using a browserify transform to inline the shaders. glslify and by extension this module can basically only be used in the browser with a build step (currently). 

Here's something you can use with webpack, it includes a glslify example in the readme. 
https://github.com/hughsk/ify-loader

**Comment by jsonnull** [comment]
I've been using https://github.com/stackgl/glslify-loader for glslify loading, but it operates at the `require()`/`import` level instead of actually inlining `glslify(...)` calls.

Properly getting the shaders imported during the webpack loader step would need the source files 'required' in order for the loader to take up the transform, not to mention ensuring the build step covered this source (typically the node_modules directory is excluded from running a project's webpack loaders).

Would `ify-loader` work the same way? I'm not sure if inlining various transforms is a typical browserify process or not. Also, would the transform have to be applied to all source, including node_modules?

**Comment by mattdesl** [comment]
glslify-loader is nice if you want to stick with the webpack require() overload syntax. It is generally only useful for local modules (ie nothing in node_modules). 

ify-loader can crawl your node_modules and find browserify config in package.json to determine which transforms to run, like brfs or glslify. With ify-loader you can require this module like normal and it should inline the shaders without any additional bloat. You can also use ify-loader to use glslify locally, using the standard browserify syntax instead of the webpack require overloads. 

So ify-loader is really all you need. 😁

Basically this problem is because bundlers cannot agree on a single format for source transforms. A nice solution would be to inline the shaders before publishing to npm but that introduces other complexities for me as an author. 

Lots more discussion on the problem here: https://gist.github.com/mattdesl/aaf759da84cc44c22305

**Comment by jsonnull** [comment]
> ify-loader can crawl your node_modules and find browserify config in package.json to determine which transforms to run

Hot damn, that's pretty nice! That's a lot better than running my project transforms across all dependencies, for sure.

Thanks for your help!
